[feature](doris-future) Support "map_entries" and "map_contains_entry" functions

DayuanX opened the pull request:
### What problem does this PR solve?
Added `map_entries` function, to return a list of struct(key, value) for each key-value pair in the map.
Added `map_contains_entry` function, to check if a map contains a given key-value pair.

When given map is null, both functions return null.

`map_contains_entry` does not support the comparison of nested type, just like `array_contains`.

Related PR: #48203 

Problem Summary:
Support "map_entries" and "map_contains_entry" functions

### Release note
Added `map_entries` function, to return a list of struct(key, value) for each key-value pair in the map.
Added `map_contains_entry` function, to check if a map contains a given key-value pair.

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [X] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [X] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [X] Yes. [link](https://github.com/apache/doris-website/pull/2779)

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

DayuanX committed:
implements map_entries

DayuanX committed:
implements map_contains_entry(map, key, value)

DayuanX committed:
add testcases

DayuanX committed:
code refactor and bugfixes

DayuanX commented:
run buildall

zclllyybb commented:
run beut

DayuanX committed:
[test](function) add combination tests

DayuanX commented:
run beut

DayuanX commented:
run check_coverage

DayuanX commented:
run buildall

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  4.88% (2/41)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_d27ba21c5d48019edc870f993eab2032b2a25c25/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_d27ba21c5d48019edc870f993eab2032b2a25c25/fe_report/index.html)

mrhhsg commented on be/src/vec/functions/function_map.cpp:
```diff
@@ -440,13 +518,409 @@ class FunctionStrToMap : public IFunction {
+
+        // extract (search) key and value columns
+        auto key_column_ptr =
+                block.get_by_position(arguments[1]).column->convert_to_full_column_if_const();
```
Maybe should try to use `unpack_if_const` instead of `convert_to_full_column_if_const`.

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `77.54% (290/374)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_d27ba21c5d48019edc870f993eab2032b2a25c25_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_d27ba21c5d48019edc870f993eab2032b2a25c25_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 81.10% (22158/27323) |
| Line Coverage     | 73.80% (229367/310789) |
| Region Coverage   | 61.63% (192362/312116) |
| Branch Coverage   | 65.32% (82676/126571) |

DayuanX committed:
[improvement](function) re-implement map_contains_entry (#54036)

Use per-column comparisons with AND to replace dual-type dispatch, avoiding code bloat

DayuanX committed:
Merge branch 'master' into add-map-functions

DayuanX committed:
[fix](function) map_contains_entry/map_entries use getFunctionParams in FE & update tests

DayuanX commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/MapContainsEntry.java:
```diff
@@ -0,0 +1,93 @@
+    @Override
+    public MapContainsEntry withChildren(List<Expression> children) {
+        Preconditions.checkArgument(children.size() == 3);
+        return new MapContainsEntry(children.get(0), children.get(1), children.get(2));
```
fixed

DayuanX commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `1.98% (7/353)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/7771a9efd03669281d2cf8e42992f7ca42ddd1a2_7771a9efd03669281d2cf8e42992f7ca42ddd1a2/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/7771a9efd03669281d2cf8e42992f7ca42ddd1a2_7771a9efd03669281d2cf8e42992f7ca42ddd1a2/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 59.56% (16784/28182) |
| Line Coverage     | 48.46% (152612/314955) |
| Region Coverage   | 37.44% (115996/309792) |
| Branch Coverage   | 40.34% (51223/126984) |

DayuanX commented:
run cloud_p0

zclllyybb commented on be/src/vec/functions/function_map.cpp:
```diff
@@ -288,6 +289,83 @@ class FunctionMapEntries : public IFunction {
+        const ColumnMap* map_column = nullptr;
+        ColumnPtr nullmap_column = nullptr;
+
+        if (left_column->is_nullable()) {
```
map_column 直接assert_cast出来就行

zclllyybb commented on be/src/vec/functions/function_map.cpp:
```diff
@@ -288,6 +289,83 @@ class FunctionMapEntries : public IFunction {
+                                                       map_column->get_offsets_ptr());
+
+        // Handle nullable case for the whole array
+        if (nullmap_column) {
```
ditto.

zclllyybb commented on be/src/vec/functions/function_map.cpp:
```diff
@@ -440,13 +518,400 @@ class FunctionStrToMap : public IFunction {
+                                     const UInt8* map_row_nullmap, bool search_is_const,
+                                     ColumnUInt8& result_matches) const {
+        switch (type) {
+        case TYPE_BOOLEAN:
```
why dont have ip types?

zclllyybb reviewed (commented):
Good job! still some comment should be resolved.

DayuanX committed:
[test](function) add map(null,null) / null test for map_contains_entry/map_entries

DayuanX committed:
[improvement][test](function) Drop handwritten comparison and support ip conparison in map_contains_entry; simplify map_entries

DayuanX commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/MapContainsEntry.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    public List<FunctionSignature> getSignatures() {
+        if (getArgument(0).getDataType().isMapType()) {
```
comments added.

DayuanX commented:
run buildall

DayuanX committed:
[style](function) code format

DayuanX commented:
run buildall

DayuanX committed:
Merge branch 'master' into add-map-functions

DayuanX commented:
run buildall

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  4.44% (2/45)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_53928cbfffb001aacb6285fe3bc01f6e7e1b9584/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_53928cbfffb001aacb6285fe3bc01f6e7e1b9584/fe_report/index.html)

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  82.22% (37/45)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_53928cbfffb001aacb6285fe3bc01f6e7e1b9584_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_53928cbfffb001aacb6285fe3bc01f6e7e1b9584_merge_fe/report/index.html)

DayuanX committed:
Merge branch 'master' into add-map-functions

DayuanX commented:
run buildall

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  4.44% (2/45)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_d0903158d7904ee98505ae7898df4b54c9fa7e07/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_d0903158d7904ee98505ae7898df4b54c9fa7e07/fe_report/index.html)

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  82.22% (37/45)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_d0903158d7904ee98505ae7898df4b54c9fa7e07_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_d0903158d7904ee98505ae7898df4b54c9fa7e07_merge_fe/report/index.html)

DayuanX commented:
run beut

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `2.93% (10/341)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/d0903158d7904ee98505ae7898df4b54c9fa7e07_d0903158d7904ee98505ae7898df4b54c9fa7e07/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/d0903158d7904ee98505ae7898df4b54c9fa7e07_d0903158d7904ee98505ae7898df4b54c9fa7e07/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 51.85% (17237/33245) |
| Line Coverage     | 37.25% (157338/422361) |
| Region Coverage   | 31.87% (119960/376351) |
| Branch Coverage   | 33.26% (52704/158454) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `85.34% (291/341)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_d0903158d7904ee98505ae7898df4b54c9fa7e07_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_d0903158d7904ee98505ae7898df4b54c9fa7e07_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 70.69% (23079/32646) |
| Line Coverage     | 57.08% (240903/422067) |
| Region Coverage   | 52.35% (199834/381763) |
| Branch Coverage   | 54.08% (86174/159348) |

DayuanX commented:
run arm

zclllyybb commented on be/src/vec/functions/function_map.cpp:
```diff
@@ -440,13 +491,409 @@ class FunctionStrToMap : public IFunction {
+        auto search_key_type = remove_nullable(block.get_by_position(arguments[1]).type);
+        auto search_value_type = remove_nullable(block.get_by_position(arguments[2]).type);
+
+        bool key_types_comparable = type_comparable(map_key_type->get_primitive_type(),
```
In FE we use `AnyDataType` with same slot or `FollowToAnyDataType`. they both will inference type to same. so I think `map_key_type->get_primitive_type()` and `search_key_type->get_primitive_type()` here should be same type?

zclllyybb reviewed (commented):
only one question still here

mrhhsg reviewed (dismissed):
lgtm

DayuanX committed:
[improvement](function) improve MapContainsEntry's type check

DayuanX committed:
Merge branch 'master' into add-map-functions

shuke987 commented:
run buildall

DayuanX commented:
run p0

DayuanX commented:
run cloud_p0

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `3.06% (10/327)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_e849abfb46e1618dfe76df052cad3865bcb19716_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_e849abfb46e1618dfe76df052cad3865bcb19716_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 57.16% (18672/32667) |
| Line Coverage     | 42.44% (179310/422499) |
| Region Coverage   | 37.06% (141824/382691) |
| Branch Coverage   | 38.12% (60868/159672) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `3.06% (10/327)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_e849abfb46e1618dfe76df052cad3865bcb19716_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_e849abfb46e1618dfe76df052cad3865bcb19716_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 57.16% (18672/32667) |
| Line Coverage     | 42.44% (179310/422499) |
| Region Coverage   | 37.06% (141824/382691) |
| Branch Coverage   | 38.12% (60868/159672) |

DayuanX committed:
[fix](test) fix map functions' testcases about duplicated keys

DayuanX commented:
run buildall

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `3.06% (10/327)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_5e839f4f289e626753530245a2bd800e0e3178c7_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_5e839f4f289e626753530245a2bd800e0e3178c7_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 57.09% (18649/32667) |
| Line Coverage     | 42.33% (178828/422496) |
| Region Coverage   | 36.85% (141028/382691) |
| Branch Coverage   | 37.96% (60613/159672) |

DayuanX committed:
[fix](function) remove incorrect asssertion

zclllyybb reviewed (approved):
LGTM

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `85.54% (278/325)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_e8d6a83347a1f716dad066b92ebaf519d4745173_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_e8d6a83347a1f716dad066b92ebaf519d4745173_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 70.88% (23160/32673) |
| Line Coverage     | 57.30% (242175/422673) |
| Region Coverage   | 52.72% (201786/382784) |
| Branch Coverage   | 54.38% (86853/159724) |

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  82.22% (37/45)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/54036_e8d6a83347a1f716dad066b92ebaf519d4745173_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/54036_e8d6a83347a1f716dad066b92ebaf519d4745173_merge_fe/report/index.html)

mrhhsg reviewed (approved):
lgtm
